core/src/main/java/io/questdb/griffin/model/WithClauseModel.java

```diff
@@ -24,36 +24,74 @@
 
 package io.questdb.griffin.model;
 
+import io.questdb.std.LowerCaseCharSequenceObjHashMap;
 import io.questdb.std.Mutable;
+import io.questdb.std.ObjList;
 import io.questdb.std.ObjectFactory;
+import org.jetbrains.annotations.Nullable;
 
 public class WithClauseModel implements Mutable {
 
     public static final ObjectFactory<WithClauseModel> FACTORY = WithClauseModel::new;
     private QueryModel model;
+    private LowerCaseCharSequenceObjHashMap<WithClauseModel> originalWithClauses;
+    private int originalWithClausesSize;
     private int position;
+    private LowerCaseCharSequenceObjHashMap<WithClauseModel> withClauses;
+
+    private boolean withClausesInitialized;
 
     private WithClauseModel() {
     }
 
     @Override
     public void clear() {
-        this.position = 0;
-        this.model = null;
+        position = 0;
+        model = null;
+        originalWithClauses = null;
+        originalWithClausesSize = -1;
+        withClauses = null;
     }
 
     public int getPosition() {
         return position;
     }
 
-    public void of(int position, QueryModel model) {
+    public LowerCaseCharSequenceObjHashMap<WithClauseModel> getWithClauses() {
+        if (!withClausesInitialized) {
+            withClauses = getSubMap();
+            withClausesInitialized = true;
+        }
+        return withClauses;
+    }
+
+    public void of(int position, LowerCaseCharSequenceObjHashMap<WithClauseModel> withClauses, QueryModel model) {
         this.position = position;
         this.model = model;
+        this.originalWithClauses = withClauses;
+        this.originalWithClausesSize = withClauses.size();
```

Comment:
nit: `size()` is cheap and can be in-place instead of storing it in a field.

Reply:
Do we compile queries on the hot path?

Reply:
size() is stored in field because withClause can grow between with clause re-evaluations.

Reply:
> size() is stored in field because withClause can grow between with clause re-evaluations.

This deserves a comment.

Reply:
Added comment in 7d0849b9 .